### PR TITLE
support ubuntu 24 in CI

### PR DESCRIPTION
**Pull request by tigercosmos**
- use `venv` for pip installation
- add Ubuntu 24.04 in the CI

### CODE REVIEW

**Comment by seladb** [comment]
@tigercosmos we need to install venv on all Linux containers to make it work. In Ubuntu the package is `python3-venv` I think, we need to figure out what the package name in the other distros

**Comment by tigercosmos** [comment]
> @tigercosmos we need to install venv on all Linux containers to make it work. In Ubuntu the package is `python3-venv` I think, we need to figure out what the package name in the other distros

I see, let me figure out.

**Comment by tigercosmos** [comment]
blocking by #1384 due to:
```
  DEPRECATION: netifaces is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559
```

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -80,6 +84,19 @@ jobs:
...
+          if [ "${{ matrix.image }}" = "alpine317" ]; then
+            apk add python3 py3-virtualenv py3-virtualenv python3-dev gcc musl-dev linux-headers
+          elif [ "${{ matrix.image }}" = "rhel94" ]; then
+            yum install -y python3 python3-virtualenv python3-devel
+          elif [ "${{ matrix.image }}" = "fedora39" ]; then
+            dnf install -y python3 python3-virtualenv python3-devel
+          else
+            apt-get update
+            apt-get install -y python3 python3-venv python3-dev
+          fi
```
Comment (clementperon) [inline .github/workflows/build_and_test.yml:98]: this should be done in the container

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -31,6 +31,7 @@ jobs:
           apk update && apk add cppcheck python3-dev
           python3 -m pip install cmake-format clang-format==18.1.6
 
+      # TODO: investigate how to run pre-commit with `venv`
       - uses: pre-commit/action@2c7b3805fd2a0fd8c1884dcaebf91fc102a13ecd # v3.0.1
```
Comment (clementperon) [inline .github/workflows/build_and_test.yml:35]: Did you try 
```suggestion
      - name: Activate virtualenv
        run: |
          ${{ matrix.python }} -m venv ./venv
          . .venv/bin/activate
          echo PATH=$PATH >> $GITHUB_ENV

      - uses: pre-commit/action@2c7b3805fd2a0fd8c1884dcaebf91fc102a13ecd # v3.0.1
```

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -52,6 +53,9 @@ jobs:
     strategy:
       matrix:
         include:
+          - image: ubuntu2404
+            python: python3
```
Comment (clementperon) [inline .github/workflows/build_and_test.yml:57]: now that everybody use python: python3 maybe we could drop it no ?

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -97,15 +114,19 @@ jobs:
...
         run: |
-          ${{ matrix.python }} -m pip install -U pip
+          ${{ matrix.python }} -m venv ./venv
+          source ./venv/bin/activate
+          ${{ matrix.python }} -m pip install pip
           ${{ matrix.python }} -m pip install -r ci/run_tests/requirements.txt
           ${{ matrix.python }} ci/run_tests/run_tests.py --interface eth0 ${{ matrix.test-flags }}
 
       - name: Test Examples
+        shell: bash
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:126]: the default shell is `sh`, and to `source` the venv activator, we need bash.

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -97,15 +114,19 @@ jobs:
         run: cmake --build "$BUILD_DIR" -j
 
       - name: Test PcapPlusPlus
+        shell: bash
         run: |
-          ${{ matrix.python }} -m pip install -U pip
+          ${{ matrix.python }} -m venv ./venv
+          source ./venv/bin/activate
+          ${{ matrix.python }} -m pip install pip
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:121]: not sure, maybe yes, because we use venv now.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:98]: ya, we can do that as well, @seladb can you update the containers?

Comment (clementperon) [reply .github/workflows/build_and_test.yml:98]: I can do it if you want, but you just need to open a MR here:
https://github.com/seladb/PcapPlusPlus-DockerImages

Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:98]: cool, I didn't know this.

Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:98]: @seladb  shouldn't [/PcapPlusPlus-DockerImages](https://github.com/seladb/PcapPlusPlus-DockerImages) under the organization?

Comment (seladb) [reply .github/workflows/build_and_test.yml:98]: I guess we can move it, but maybe do it after the release?

File: .github/workflows/build_and_test.yml (hunk 2 above)
Comment (clementperon) [reply .github/workflows/build_and_test.yml:40]: @tigercosmos it shoul be "source ./venv/bin/activate"

Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:40]: `source` is only for bash.  However, I cannot control to let pre-commit action to use bash. I will just convert the change, and I don't see any issue that pre-commit doesn't run in the venv.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:98]: got it.

File: .github/workflows/build_and_test.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml]: fixed.

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml]: removed all.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml]: fixed.

**Comment by tigercosmos** [comment]
@seladb  ready for review.

File: .github/workflows/build_and_test.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (seladb) [reply .github/workflows/build_and_test.yml:117]: In `sh`, instead of using `source` you can use `.`:
```
. ./venv/bin/activate
```

https://unix.stackexchange.com/questions/748962/how-to-activate-python-venv-in-a-non-bash-shell

[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:117]: I tried, but it doesn't work.

[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml]: okay, let me try again.

[the remark this answers is not in the document]
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml]: ah, well, seems it works :)

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -32,6 +32,7 @@ jobs:
           apk update && apk add cppcheck python3-dev
           python3 -m pip install cmake-format clang-format==18.1.6
 
+      # TODO: investigate how to run pre-commit with `venv`
```
Comment (egecetin) [inline .github/workflows/build_and_test.yml:35]: Not related with this PR directly but, it is possible to use [pre-commit.ci](https://https://pre-commit.ci/) but we should remove cppcheck and clang-format from pre-commit since they require pre-installed binaries.

**Review by clementperon — APPROVED** [review]

File: .github/workflows/build_and_test.yml [hunk 7]
```diff
@@ -111,16 +107,17 @@ jobs:
 
       - name: Test PcapPlusPlus
         run: |
-          ${{ matrix.python }} -m pip install -U pip
-          ${{ matrix.python }} -m pip install -r ci/run_tests/requirements.txt
-          ${{ matrix.python }} ci/run_tests/run_tests.py --interface eth0 ${{ matrix.test-flags }}
+          python3 -m venv ./venv
```
Comment (Dimi1010) [inline .github/workflows/build_and_test.yml:110]: nit: Maybe move this line to it's own step? It looks weird to me that the venv is created inside the 'Test PCPP' step but the following steps also depend on it.

File: .github/workflows/build_and_test.yml [hunk 8]
```diff
@@ -347,15 +345,15 @@ jobs:
...
           python -m pip install -r ci/run_tests/requirements.txt
           python ci/run_tests/run_tests.py --interface en0
 
       - name: Test Examples
         if: ${{ matrix.host-arch == matrix.arch }}
         run: |
           cd Tests/ExamplesTest
-          python -m pip install -U pip
           python -m pip install -r requirements.txt
           python -m pytest --interface en0 --use-sudo --root-path=../../Dist/examples_bin
```
Comment (Dimi1010) [inline .github/workflows/build_and_test.yml:358]: Why does the first step use 'venv' and the second doesn't?

**Review by seladb — APPROVED** [review]
Please fix @Dimi1010 's comments, otherwise LGTM

File: .github/workflows/build_and_test.yml (hunk 7 above)
Comment (tigercosmos) [reply .github/workflows/build_and_test.yml:110]: Sure. Moved it to a step.
